### DISCUSSION TITLE
How to get asset from dictionary on "Bingx" in Python

### DISCUSSION THREAD

**Discussion by allg97** [amirinsight/py-bingx#2] [Q&A] [answered]
I wrote "def" that it fetch contracts like below:
```

instruments =send_request(.......)
contractdict = dict()        
if instruments is not None:
   for s in instruments['data']:          
       contractdict [s['symbol']] = Contract(s, "bingx")
```

And it works correctly.

Now I write another "def" like this, but I want to fetch 'asset', I put it below:
```

margindata =send_request(.......)
balancedic = dict()
if margindata is not None:            
   for b in margindata ['balance']:
       balancedic [b['asset']] = balance(b, "bingx")
```

But, it has type error!

>     TypeError: string indices must be integers, not 'str'

**Send_request** in both the above functions will return the correct answer.

I also put the reference link below:

[](https://bingx-api.github.io/docs/#/swapV2/account-api.html#Get%20Perpetual%20Swap%20Account%20Asset%20Information)

I also test to change **'key'** name such as **'asset','data','balance'**, but they are wrong too. Please tell me how to fetch every fields of **margindata** dictionary like instruments. 

***I not used any other library like ccxt,... . I just used this.

**Comment by amirinsight** [comment]
What is the contents of "margindata"? Which endpoint are you using?

**Reply by allg97** [reply]
my end point is : '/openApi/swap/v2/user/balance'
and the margin data contetnt is: 
{'code': 0, 'msg': '', 'data': {'balance': {...}}}

Also 'data' is :{'balance': {'userId': 'myuserID', 'asset': 'USDT', 'balance': '0.0000', 'equity': '0.0000', 'unrealizedProfit': '0.0000', 'realisedProfit': '0.0000', 'availableMargin': '0.0000', 'usedMargin': '0.0000', 'freezedMargin': '0.0000'}}

**Reply by amirinsight** [reply]
First debug your code to check which one these lines gives this error.

Is it in `for b in margindata ['balance']:`
 Or is it this line? `balancedic [b['asset']] = balance(b, "bingx")`

Also make sure `margindata` is a json object.

**Reply by allg97** [reply]
These lines works correctly. when it goes into --_balance(b, "bingx")_-- **destination class** to save the fields one by one, an error occurs. like this: self.balance = float(info['balance']). 

![indice integer](https://github.com/amirinsight/py-bingx/assets/144830546/487da2e2-e1fc-48cd-a60f-bfda4aa1d738)

**Reply by amirinsight** [reply, marked as answer]
So what's happening here is that "b" that is being given to the "Balance" class is a string and it is not a json object. So your class cannot initialize correctly.

Here is how you can fix this:
```python
from bingx.api import BingxAPI

bingx = BingxAPI(API_KEY, SECRET_KEY, timestamp="local")

class Balance(dict):
    """
    Returns a python dict object
    """
    def __init__(self, data, exchange):
        if exchange == "bingx":
            balance_data = data['data']['balance']
            my_dict = {
                'balance': float(balance_data['balance']),
                'equity': float(balance_data['equity']),
                'unrealizedProfit': float(balance_data['unrealizedProfit']),
                'realizedProfit': float(balance_data['realisedProfit']),
                'availableMargin': float(balance_data['availableMargin']),
                'usedMargin': float(balance_data['usedMargin']),
                'freezedMargin': float(balance_data['freezedMargin'])
            }
            super().__init__(my_dict)


data = bingx.get_perpetual_balance()
if data is not None:
    balancedic = Balance(data, "bingx")
    print(balancedic)
    print(balancedic['balance'])
```

Output:

```
{'balance': '0.0000', 'equity': '0.0000', 'unrealizedProfit': '0.0000', 'realizedProfit': '0.0000', 'availableMargin': '0.0000', 'usedMargin': '0.0000', 'freezedMargin': '0.0000'}
0.0000
```

Hope it helps!

**Comment by amirinsight** [comment]
@allg97 Is your problem solved?
